File: host-libs/js/scanner.ts
import * as path from "https://deno.land/std@0.208.0/path/mod.ts";

export interface Manifest {
  bundle_name: string;
  runtime: string;
  version: string;
  file: string;
  provides: string[];
  function_count: Record<string, number>;
}

export interface Bundle {
  path: string;
  manifest: Manifest;
}

export function scanDir(dirPath: string): Bundle[] {
  const bundles: Bundle[] = [];
  
  for (const entry of Deno.readDirSync(dirPath)) {
    if (!entry.isDirectory) continue;
    
    const manifestPath = path.join(dirPath, entry.name, "manifest.toml");
    try {
      const content = Deno.readTextFileSync(manifestPath);
      const manifest = parseToml(content) as Manifest;
      bundles.push({
        path: path.join(dirPath, entry.name),
        manifest,
      });
    } catch {
      // No manifest.toml in this directory
    }
  }
  
  return bundles;
}

// Simple TOML parser for manifest files
function parseToml(content: string): Record<string, any> {
  const result: Record<string, any> = {};
  let currentSection: string | null = null;
  
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#')) continue;
    
    // Section header [section] or [section.subsection]
    const sectionMatch = trimmed.match(/^\[(.+)\]$/);
    if (sectionMatch) {
      currentSection = sectionMatch[1];
      if (!result[currentSection]) {
        result[currentSection] = {};
      }
      continue;
    }
    
    // Key-value pair
    const kvMatch = trimmed.match(/^(\w+)\s*=\s*(.+)$/);
    if (kvMatch) {
      const key = kvMatch[1];
      let value = kvMatch[2].trim();
      
      // Parse value
      if (value.startsWith('"') && value.endsWith('"')) {
        value = value.slice(1, -1);
      } else if (value.startsWith('[') && value.endsWith(']')) {
        // Array
        value = value.slice(1, -1).split(',').map(s => s.trim().replace(/^"|"$/g, ''));
      } else if (/^\d+$/.test(value)) {
        value = parseInt(value, 10);
      }
      
      if (currentSection) {
        result[currentSection][key] = value;
      } else {
        result[key] = value;
      }
    }
  }
  
  return result;
}

File: host-libs/lua/scanner.lua
-- Scanner module for discovering polyplug plugin bundles

local ffi = require('ffi')
local M = {}

local function parse_toml(content)
    local result = {}
    local current_section = nil
    
    for line in content:gmatch('[^\n]+') do
        line = line:gsub('^%s+', ''):gsub('%s+$', '')
        if #line > 0 and not line:match('^#') then
            -- Section header
            local section = line:match('^%[(.+)%]$')
            if section then
                current_section = section
                result[current_section] = {}
            else
                -- Key-value pair
                local key, value = line:match('^(%w+)%s*=%s*(.+)$')
                if key and value then
                    value = value:gsub('^"', ''):gsub('"$', '')
                    if value:match('^%[') then
                        -- Array
                        local arr = {}
                        for item in value:gmatch('"([^"]+)"') do
                            table.insert(arr, item)
                        end
                        value = arr
                    elseif value:match('^%d+$') then
                        value = tonumber(value)
                    end
                    
                    if current_section then
                        result[current_section][key] = value
                    else
                        result[key] = value
                    end
                end
            end
        end
    end
    
    return result
end

function M.scan_dir(dir_path)
    local bundles = {}
    
    local dir = io.popen('find "' .. dir_path .. '" -maxdepth 1 -type d -mindepth 1')
    for subdir in dir:lines() do
        local manifest_path = subdir .. '/manifest.toml'
        local file = io.open(manifest_path, 'r')
        if file then
            local content = file:read('*all')
            file:close()
            local manifest = parse_toml(content)
            table.insert(bundles, { path = subdir, manifest = manifest })
        end
    end
    dir:close()
    
    return bundles
end

return M
